src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,792 @@
+        }
+
+        private sealed class Converter
+        {
+            private readonly SemanticModel _semanticModel;
+            private readonly ISemanticFactsService _semanticFacts;
+            private readonly CancellationToken _cancellationToken;
+
+            public Converter(SemanticModel semanticModel, ISemanticFactsService semanticFacts, CancellationToken cancellationToken)
+            {
+                _semanticModel = semanticModel;
+                _semanticFacts = semanticFacts;
+                _cancellationToken = cancellationToken;
+            }
+
+            public bool TryConvert(QueryExpressionSyntax source, out DocumentUpdateInfo documentUpdateInfo)
+            {
+                if (source.DescendantTrivia().Any(trivia => trivia.MatchesKind(
+                        SyntaxKind.SingleLineCommentTrivia,
+                        SyntaxKind.MultiLineCommentTrivia,
+                        SyntaxKind.MultiLineDocumentationCommentTrivia) ||
+                    source.ContainsDirectives))
+                {
+                    documentUpdateInfo = default;
+                    return false;
+                }
+
+                // GetDiagnostics is expensive. Move it to the end if there were no bail outs from the algorithm.
+                // TODO likely adding more semantic checks will perform checks we expect from GetDiagnostics
+                // We may consider removing GetDiagnostics.
+                // https://github.com/dotnet/roslyn/issues/25639
+                if (!TryConvertInternal(source, out documentUpdateInfo) ||
+                    _semanticModel.GetDiagnostics(source.Span, _cancellationToken).Any())
+                {
+                    documentUpdateInfo = default;
+                    return false;
+                }
+
+                return true;
+            }
+
+            private bool TryConvertInternal(QueryExpressionSyntax source, out DocumentUpdateInfo documentUpdateInfo)
+            {
+                // (from a in b select a); 
+                SyntaxNode parent = source.WalkUpParentheses().Parent;
+
+                switch (parent.Kind())
+                {
+                    // return from a in b select a;
+                    case SyntaxKind.ReturnStatement:
+                        return TryConvertIfInReturnStatement(source, (ReturnStatementSyntax)parent, out documentUpdateInfo);
+                    // foreach(var x in from a in b select a)
+                    case SyntaxKind.ForEachStatement:
+                        return TryConvertIfInForEach(source, (ForEachStatementSyntax)parent, out documentUpdateInfo);
+                    // (from a in b select a).ToList(), (from a in b select a).Count(), etc.
+                    case SyntaxKind.SimpleMemberAccessExpression:
+                        return TryConvertIfInMemberAccessExpression(source, (MemberAccessExpressionSyntax)parent, out documentUpdateInfo);
```

Comment:
>TryConvertIfInMemberAccessExpression [](start = 31, length = 36)

Should we try TryReplaceWithLocalFunction if TryConvertIfInMemberAccessExpression fails?